### PR TITLE
Make `allowrankdeficient` a keyword argument

### PR DESCRIPTION
**Pull request by pdeffebach**
In reference to #351. 

Whether or not `lm` should allow a rank deficient matrix is an option, and `true` or `false` is not an input value on its own. Therefore it makes sense that this be a keyword argument and not a positional argument.

The docs are also improved. 

No changes to `glm` because I don't think any aspect of `glm` allows rank deficient input matrices, even though it probably should. But thats another PR.

### CODE REVIEW

**Comment by pdeffebach** [comment]
Tests fail for unrelated reasons. I will try and figure out and fix with another PR. 

This is ready for a review.

File: src/lm.jl [hunk 1]
```diff
@@ -140,9 +147,14 @@ to its weight. Do note that this interpretation gives equal point estimates but
...
+
+`allowrankdeficient` controls whether or not `lm` accepts a model matrix which
+is less-than-full rank. If `true`, only the first linearly-dependent columns are used.
+The coefficient on redundant linearly dependent columns is `0.0` and all other
+associated statistics are set to `NaN`.
 """
-lm(X, y, allowrankdeficient::Bool=false; kwargs...) = 
-    fit(LinearModel, X, y, allowrankdeficient; kwargs...)
+lm(X, y, allowrankdeficient_dep::Union{Bool,Nothing}=nothing; kwargs...) =
+    fit(LinearModel, X, y, nothing; kwargs...)
```
Comment (pdeffebach) [inline src/lm.jl:157]:
```suggestion
    fit(LinearModel, X, y, allowrankdeficient_dep; kwargs...)
```

**Comment by pdeffebach** [comment]
Tests pass now, the "trim trailing whitespace" option in Sublime text was breaking the printing tests. 

Ready for a review.

File: test/runtests.jl [hunk 2]
```diff
@@ -84,13 +84,21 @@ end
...
-    m2p = fit(LinearModel, Xmissingcell, ymissingcell, true)
+    m2p = fit(LinearModel, Xmissingcell, ymissingcell; allowrankdeficient = true)
     @test isa(m2p.pp.chol, CholeskyPivoted)
     @test rank(m2p.pp.chol) == 11
     @test isapprox(deviance(m2p), 0.2859221258731563)
     @test isapprox(coef(m2p), [0.9178241203127236, 9.089883493902754, 3.01742566831296,
                    4.108734932819495, 4.995249696954908, 6.075962907632594, 0.0, 8.038151489191618,
                    8.848886704358202, 2.8697881579099085, 11.15107375630744, 11.8392578374927])
+
+    m2p_dep = fit(LinearModel, Xmissingcell, ymissingcell, true)
```
Comment (kleinschmidt) [inline test/runtests.jl:95]: maybe add a test that this `warn`s?

Comment (kleinschmidt) [reply test/runtests.jl:95]: and I guess maybe also the behavior when the kwarg and the positional conflict?

File: src/glmfit.jl [hunk 3]
```diff
@@ -58,7 +58,7 @@ function GlmResp(y::V, d::D, l::L, off::V, wts::V) where {V<:FPVector,D,L}
     return r
 end
 
-function GlmResp(y::AbstractVector{<:Real}, d::D, l::L, off::AbstractVector{<:Real}, 
+function GlmResp(y::AbstractVector{<:Real}, d::D, l::L, off::AbstractVector{<:Real},
```
Comment (kleinschmidt) [inline src/glmfit.jl:61]: can you revert all these whitespace changes so that the git blame is clean?

**Review by kleinschmidt — CHANGES_REQUESTED** [review]
LGTM, other than a few comments on the tests

File: src/lm.jl [hunk 4]
```diff
@@ -114,23 +114,30 @@ LinearAlgebra.cholesky(x::LinearModel) = cholesky(x.pp)
...
 
 function fit(::Type{LinearModel}, X::AbstractMatrix{<:Real}, y::AbstractVector{<:Real},
-             allowrankdeficient::Bool=false; wts::AbstractVector{<:Real}=similar(y, 0))
+             allowrankdeficient_dep::Union{Bool,Nothing}=nothing;
+             wts::AbstractVector{<:Real}=similar(y, 0),
+             allowrankdeficient::Bool=false)
+    if allowrankdeficient_dep != nothing
+        @warn "Positional argument `allowrankdeficient` is deprecated, use keyword" *
+              "argument instead. Proceeding with positional argument  value: $allowrankdeficient_dep"
+              allowrankdeficient = allowrankdeficient_dep
```
Comment (nalimilan) [inline src/lm.jl:132]:
```suggestion
    if allowrankdeficient_dep != nothing
        Base.depwarn("Positional argument `allowrankdeficient` is deprecated, use keyword" *
                     "argument instead. Proceeding with positional argument  value: $allowrankdeficient_dep",
                     :fit)
              allowrankdeficient = allowrankdeficient_dep
```

Comment (pdeffebach) [reply src/lm.jl:132]: I couldn't get this to work, I'll try again in a bit.

File: src/lm.jl [hunk 5]
```diff
@@ -117,20 +117,27 @@ function StatsBase.fit!(obj::LinearModel)
     else 
         delbeta!(obj.pp, obj.rr.y, obj.rr.wts)
     end
-    installbeta!(obj.pp)     
+    installbeta!(obj.pp)
```
Comment (pdeffebach) [inline src/lm.jl:120]:
```suggestion
    installbeta!(obj.pp)     
```

File: test/runtests.jl (hunk 2 above)
Comment (pdeffebach) [reply test/runtests.jl]: Added!

File: src/glmfit.jl (hunk 3 above)
Comment (pdeffebach) [reply src/glmfit.jl]: Fixed.

File: src/lm.jl (hunk 4 above)
Comment (pdeffebach) [reply src/lm.jl]: Got it to work!

**Comment by pdeffebach** [comment]
Thanks for the reviews. I responded to all comments. 

Ready for another round!

**Review by kleinschmidt — APPROVED** [review]

File: src/lm.jl [hunk 6]
```diff
@@ -123,14 +123,21 @@ function StatsBase.fit!(obj::LinearModel)
 end
 
 function fit(::Type{LinearModel}, X::AbstractMatrix{<:Real}, y::AbstractVector{<:Real},
-             allowrankdeficient::Bool=false; wts::AbstractVector{<:Real}=similar(y, 0))
+             allowrankdeficient_dep::Union{Bool,Nothing}=nothing;
+             wts::AbstractVector{<:Real}=similar(y, 0),
+             allowrankdeficient::Bool=false)
+    if allowrankdeficient_dep != nothing
```
Comment (nalimilan) [inline src/lm.jl:129]: This helps the compiler to completely drop this code when the argument is not `nothing`.
```suggestion
    if allowrankdeficient_dep !== nothing
```

File: src/lm.jl [hunk 7]
```diff
@@ -123,14 +123,21 @@ function StatsBase.fit!(obj::LinearModel)
...
+        @warn "Positional argument `allowrankdeficient` is deprecated, use keyword" *
+                    "argument instead. Proceeding with positional argument  value: $allowrankdeficient_dep"
```
Comment (nalimilan) [inline src/lm.jl:131]:
```suggestion
        @warn "Positional argument `allowrankdeficient` is deprecated, use keyword " *
              "argument instead. Proceeding with positional argument value: $allowrankdeficient_dep"
```

File: src/lm.jl [hunk 8]
```diff
@@ -140,9 +147,14 @@ to its weight. Do note that this interpretation gives equal point estimates but
 different standard errors from analytical (a.k.a. inverse variance) weights and
 from probability (a.k.a. sampling) weights which are the default in some other
 software.
+
+`allowrankdeficient` controls whether or not `lm` accepts a model matrix which
+is less-than-full rank. If `true`, only the first linearly-dependent columns are used.
+The coefficient on redundant linearly dependent columns is `0.0` and all other
```
Comment (nalimilan) [inline src/lm.jl:153]:
```suggestion
The coefficient on redundant linearly dependent columns is `0.0` and all
```

File: test/runtests.jl [hunk 9]
```diff
@@ -84,13 +84,26 @@ end
...
+    m2p = fit(LinearModel, Xmissingcell, ymissingcell; allowrankdeficient = true)
     @test isa(m2p.pp.chol, CholeskyPivoted)
     @test rank(m2p.pp.chol) == 11
     @test isapprox(deviance(m2p), 0.2859221258731563)
     @test isapprox(coef(m2p), [0.9178241203127236, 9.089883493902754, 3.01742566831296,
                    4.108734932819495, 4.995249696954908, 6.075962907632594, 0.0, 8.038151489191618,
                    8.848886704358202, 2.8697881579099085, 11.15107375630744, 11.8392578374927])
+
+    m2p_dep_pos = fit(LinearModel, Xmissingcell, ymissingcell, true)
+    @test_logs (:warn, "Positional argument `allowrankdeficient` is deprecated, use keywordargument instead. Proceeding with positional argument  value: true") fit(LinearModel, Xmissingcell, ymissingcell, true)
```
Comment (nalimilan) [inline test/runtests.jl:96]:
```suggestion
    @test_logs (:warn, "Positional argument `allowrankdeficient` is deprecated, use keyword argument instead. " *
                "Proceeding with positional argument value: true") fit(LinearModel, Xmissingcell, ymissingcell, true)
```

File: test/runtests.jl [hunk 10]
```diff
@@ -84,13 +84,26 @@ end
...
+    @test isa(m2p_dep_pos.pp.chol, CholeskyPivoted)
+    @test rank(m2p_dep_pos.pp.chol) == rank(m2p.pp.chol)
+    @test isapprox(deviance(m2p_dep_pos),deviance(m2p))
+    @test isapprox(coef(m2p_dep_pos), coef(m2p))
+
+    m2p_dep_pos_kw = fit(LinearModel, Xmissingcell, ymissingcell, true; allowrankdeficient = false)
+    @test isa(m2p_dep_pos_kw.pp.chol, CholeskyPivoted)
+    @test rank(m2p_dep_pos_kw.pp.chol) == rank(m2p.pp.chol)
+    @test isapprox(deviance(m2p_dep_pos_kw),deviance(m2p))
```
Comment (nalimilan) [inline test/runtests.jl:105]:
```suggestion
    @test isapprox(deviance(m2p_dep_pos_kw), deviance(m2p))
```

File: test/runtests.jl [hunk 11]
```diff
@@ -84,13 +84,26 @@ end
...
     @test isapprox(deviance(m2p), 0.2859221258731563)
     @test isapprox(coef(m2p), [0.9178241203127236, 9.089883493902754, 3.01742566831296,
                    4.108734932819495, 4.995249696954908, 6.075962907632594, 0.0, 8.038151489191618,
                    8.848886704358202, 2.8697881579099085, 11.15107375630744, 11.8392578374927])
+
+    m2p_dep_pos = fit(LinearModel, Xmissingcell, ymissingcell, true)
+    @test_logs (:warn, "Positional argument `allowrankdeficient` is deprecated, use keywordargument instead. Proceeding with positional argument  value: true") fit(LinearModel, Xmissingcell, ymissingcell, true)
+    @test isa(m2p_dep_pos.pp.chol, CholeskyPivoted)
+    @test rank(m2p_dep_pos.pp.chol) == rank(m2p.pp.chol)
+    @test isapprox(deviance(m2p_dep_pos),deviance(m2p))
```
Comment (nalimilan) [inline test/runtests.jl:99]:
```suggestion
    @test isapprox(deviance(m2p_dep_pos), deviance(m2p))
```

**Comment by nalimilan** [comment]
By making this a keyword argument, the name `allowrankdeficient` becomes part of the API, so I think it's worth considering whether it's the best possible name. I must say I don't find it very user-friendly, but I don't have very good suggestions either. Are there examples of similar options in other software?

**Comment by pdeffebach** [comment]
I also don't have any good suggestions. Stata and R do it by default, while it looks like python doesn't allow it at all. 

Would it be too breaking to make it `true` by default?

**Comment by nalimilan** [comment]
AFAICT it wouldn't be breaking at all since it would turn an error into something that succeeds.

@dmbates @andreasnoack Opinions?

**Comment by pdeffebach** [comment]
Okay just pushed a commit to 

1. Rename `allowrankdeficient` to `dropcollinear`
2. Make the default of `collinear` to `true`. This will cause fewer "what's going on?" scenarios for new users who are less familiar with the math behind OLS. 

It also has an added performance benefit! 

```
julia> x = rand(10000, 50);

julia> @btime GLM.cholpred(x, true);
  2.134 ms (17 allocations: 7.67 MiB)

julia> @btime GLM.cholpred(x, true);
  2.133 ms (17 allocations: 7.67 MiB)

julia> @btime GLM.cholpred($x, true);
  2.130 ms (17 allocations: 7.67 MiB)

julia> @btime GLM.cholpred($x, false);
  3.031 ms (15 allocations: 7.67 MiB)

julia> @btime GLM.cholpred($x, false);
  3.029 ms (15 allocations: 7.67 MiB)
```

Tests are updated. I haven't added anything to `docs/` as I'm not sure where it will go. But the docstring is updated.

File: src/lm.jl [hunk 12]
```diff
@@ -28,10 +28,10 @@ mutable struct LmResp{V<:FPVector} <: ModResp  # response in a linear model
     end
 end
 
-LmResp(y::FPVector, wts::FPVector=similar(y, 0)) = 
+LmResp(y::FPVector, wts::FPVector=similar(y, 0)) =
```
Comment (pdeffebach) [inline src/lm.jl:31]:
```suggestion
LmResp(y::FPVector, wts::FPVector=similar(y, 0)) = 
```

File: src/lm.jl [hunk 13]
```diff
@@ -65,7 +65,7 @@ function nulldeviance(r::LmResp)
     wts = r.wts
     if isempty(wts)
         m = mean(y)
-    else 
+    else
```
Comment (pdeffebach) [inline src/lm.jl:68]:
```suggestion
    else 
```

File: src/lm.jl [hunk 14]
```diff
@@ -28,10 +28,10 @@ mutable struct LmResp{V<:FPVector} <: ModResp  # response in a linear model
     end
 end
 
-LmResp(y::FPVector, wts::FPVector=similar(y, 0)) = 
+LmResp(y::FPVector, wts::FPVector=similar(y, 0)) =
     LmResp{typeof(y)}(fill!(similar(y), 0), similar(y, 0), wts, y)
 
-LmResp(y::AbstractVector{<:Real}, wts::AbstractVector{<:Real}=similar(y, 0)) = 
+LmResp(y::AbstractVector{<:Real}, wts::AbstractVector{<:Real}=similar(y, 0)) =
```
Comment (pdeffebach) [inline src/lm.jl:34]:
```suggestion
LmResp(y::AbstractVector{<:Real}, wts::AbstractVector{<:Real}=similar(y, 0)) = 
```

File: src/lm.jl [hunk 15]
```diff
@@ -89,7 +89,7 @@ end
 
 function nullloglikelihood(r::LmResp)
     n = isempty(r.wts) ? length(r.y) : sum(r.wts)
-    -n/2 * (log(2π * nulldeviance(r)/n) + 1) 
+    -n/2 * (log(2π * nulldeviance(r)/n) + 1)
```
Comment (pdeffebach) [inline src/lm.jl:92]:
```suggestion
    -n/2 * (log(2π * nulldeviance(r)/n) + 1) 
```

File: src/lm.jl [hunk 16]
```diff
@@ -114,23 +114,30 @@ LinearAlgebra.cholesky(x::LinearModel) = cholesky(x.pp)
 function StatsBase.fit!(obj::LinearModel)
     if isempty(obj.rr.wts)
         delbeta!(obj.pp, obj.rr.y)
-    else 
+    else
```
Comment (pdeffebach) [inline src/lm.jl:117]:
```suggestion
    else 
```

File: src/lm.jl [hunk 17]
```diff
@@ -114,23 +114,30 @@ LinearAlgebra.cholesky(x::LinearModel) = cholesky(x.pp)
...
         delbeta!(obj.pp, obj.rr.y, obj.rr.wts)
     end
-    installbeta!(obj.pp)     
+    installbeta!(obj.pp)
```
Comment (pdeffebach) [inline src/lm.jl:120]:
```suggestion
    installbeta!(obj.pp)    
```

File: test/runtests.jl [hunk 18]
```diff
@@ -49,7 +49,7 @@ linreg(x::AbstractVecOrMat, y::AbstractVector) = qr!(simplemm(x)) \ y
     @test coef(lm3) == coef(lm4) ≈ [11, 1, 2, 3, 4]
 end
 
-@testset "linear model with weights" begin 
+@testset "linear model with weights" begin
```
Comment (pdeffebach) [inline test/runtests.jl:52]:
```suggestion
@testset "linear model with weights" begin 
```

File: test/runtests.jl [hunk 19]
```diff
@@ -530,7 +544,7 @@ end
         ─────────────────────────────────────────────────────────────────
              DOF  ΔDOF     SSR    ΔSSR       R²      ΔR²        F*  p(>F)
         ─────────────────────────────────────────────────────────────────
-        [1]    3        0.1283           0.9603                          
+        [1]    3        0.1283           0.9603
```
Comment (pdeffebach) [inline test/runtests.jl:547]:
```suggestion
        [1]    3        0.1283           0.9603                          
```

File: test/runtests.jl [hunk 20]
```diff
@@ -542,7 +556,7 @@ end
         ─────────────────────────────────────────────────────────────────
              DOF  ΔDOF     SSR     ΔSSR       R²     ΔR²        F*  p(>F)
         ─────────────────────────────────────────────────────────────────
-        [1]    2        3.2292           -0.0000                         
+        [1]    2        3.2292           -0.0000
```
Comment (pdeffebach) [inline test/runtests.jl:559]:
```suggestion
        [1]    2        3.2292           -0.0000                         
```

File: test/runtests.jl [hunk 21]
```diff
@@ -556,7 +570,7 @@ end
         ──────────────────────────────────────────────────────────────────
              DOF  ΔDOF     SSR     ΔSSR       R²     ΔR²        F*   p(>F)
         ──────────────────────────────────────────────────────────────────
-        [1]    2        3.2292           -0.0000                          
+        [1]    2        3.2292           -0.0000
```
Comment (pdeffebach) [inline test/runtests.jl:573]:
```suggestion
        [1]    2        3.2292           -0.0000                          
```

File: test/runtests.jl [hunk 22]
```diff
@@ -570,7 +584,7 @@ end
         ──────────────────────────────────────────────────────────────────
              DOF  ΔDOF     SSR    ΔSSR       R²      ΔR²        F*   p(>F)
         ──────────────────────────────────────────────────────────────────
-        [1]    5        0.1017           0.9685                           
+        [1]    5        0.1017           0.9685
```
Comment (pdeffebach) [inline test/runtests.jl:587]:
```suggestion
        [1]    5        0.1017           0.9685                           
```

**Comment by pdeffebach** [comment]
Ready for a review!

File: test/runtests.jl [hunk 23]
```diff
@@ -61,13 +61,13 @@ end
     @test isapprox(stderror(lm_model), [9.382302620120193, 0.00816741377772968])
     @test isapprox(r2(lm_model), 0.8330258148644486)
     @test isapprox(adjr2(lm_model), 0.832788298242634)
-    @test isapprox(vcov(lm_model), [88.02760245551447 -0.06772589439264813; 
+    @test isapprox(vcov(lm_model), [88.02760245551447 -0.06772589439264813;
```
Comment (pdeffebach) [inline test/runtests.jl:64]:
```suggestion
    @test isapprox(vcov(lm_model), [88.02760245551447 -0.06772589439264813; 
```

File: test/runtests.jl [hunk 24]
```diff
@@ -61,13 +61,13 @@ end
...
                                     -0.06772589439264813 6.670664781664879e-5])
     @test isapprox(first(predict(lm_model)), 357.57694841780994)
     @test isapprox(loglikelihood(lm_model), -4353.946729075838)
     @test isapprox(loglikelihood(glm_model), -4353.946729075838)
     @test isapprox(nullloglikelihood(lm_model), -4984.892139711452)
-    @test isapprox(mean(residuals(lm_model)), -5.412966629787718) 
+    @test isapprox(mean(residuals(lm_model)), -5.412966629787718)
```
Comment (pdeffebach) [inline test/runtests.jl:70]:
```suggestion
    @test isapprox(mean(residuals(lm_model)), -5.412966629787718) 
```

File: src/lm.jl [hunk 25]
```diff
@@ -117,20 +117,27 @@ function StatsBase.fit!(obj::LinearModel)
     else 
         delbeta!(obj.pp, obj.rr.y, obj.rr.wts)
     end
-    installbeta!(obj.pp)     
+    installbeta!(obj.pp)
```
Comment (pdeffebach) [inline src/lm.jl:120]:
```suggestion
    installbeta!(obj.pp)     
```

**Comment by pdeffebach** [comment]
Bumping this! Just fixed some errant white space changes. 

To recap: Changes `allowrankdeficient` to a keyword argument and the new name is `dropcollinear`

**Review by kleinschmidt — APPROVED** [review]

**Comment by kleinschmidt** [comment]
LGTM.  Nightly failures are due to the p-value printing thing with printfpocalypse (so unrelated)

File: src/lm.jl [hunk 26]
```diff
@@ -140,9 +147,14 @@ to its weight. Do note that this interpretation gives equal point estimates but
 different standard errors from analytical (a.k.a. inverse variance) weights and
 from probability (a.k.a. sampling) weights which are the default in some other
 software.
+
+`dropcollinear` controls whether or not `lm` accepts a model matrix which
+is less-than-full rank. If `true`, the default, only the first linearly-dependent
+columns are used. The coefficient on redundant linearly dependent columns is
```
Comment (nalimilan) [inline src/lm.jl:153]:
```suggestion
is less-than-full rank. If `true` (the default), only the first of each set of
linearly-dependent columns is used. The coefficient for redundant linearly dependent columns is
```

**Review by nalimilan — APPROVED** [review]
Looks good! Just a wording suggestion.

**Review by pdeffebach — COMMENTED** [review]
Thanks!

**Pull request merged**